llvm/include/llvm/Support/Mustache.h

```diff
@@ -0,0 +1,190 @@
+
+namespace llvm {
+namespace mustache {
+
+using Accessor = SmallVector<SmallString<128>>;
+using Lambda = std::function<llvm::json::Value()>;
+using SectionLambda = std::function<llvm::json::Value(StringRef)>;
+
+class Token {
+public:
+  enum class Type {
+    Text,
+    Variable,
+    Partial,
+    SectionOpen,
+    SectionClose,
+    InvertSectionOpen,
+    UnescapeVariable,
+    Comment,
+  };
+
+  Token(StringRef Str);
+
+  Token(StringRef RawBody, StringRef Str, char Identifier);
+
+  StringRef getTokenBody() const { return TokenBody; };
+
+  StringRef getRawBody() const { return RawBody; };
+
+  void setTokenBody(SmallString<128> NewBody) { TokenBody = NewBody; };
+
+  Accessor getAccessor() const { return Accessor; };
+
+  Type getType() const { return TokenType; };
+
+  static Type getTokenType(char Identifier);
+
+private:
+  Type TokenType;
+  SmallString<128> RawBody;
+  Accessor Accessor;
+  SmallString<128> TokenBody;
+};
+
+class ASTNode {
+public:
+  enum Type {
+    Root,
+    Text,
+    Partial,
+    Variable,
+    UnescapeVariable,
+    Section,
+    InvertSection,
+  };
+
+  ASTNode() : T(Type::Root), LocalContext(nullptr) {};
+
+  ASTNode(StringRef Body, ASTNode *Parent)
+      : T(Type::Text), Body(Body), Parent(Parent), LocalContext(nullptr) {};
+
+  // Constructor for Section/InvertSection/Variable/UnescapeVariable
+  ASTNode(Type T, Accessor Accessor, ASTNode *Parent)
+      : T(T), Parent(Parent), Children({}), Accessor(Accessor),
+        LocalContext(nullptr) {};
+
+  void addChild(ASTNode *Child) { Children.emplace_back(Child); };
+
+  SmallString<128> getBody() const { return Body; };
+
+  void setBody(StringRef NewBody) { Body = NewBody; };
+
+  void setRawBody(StringRef NewBody) { RawBody = NewBody; };
+
+  SmallString<128> render(llvm::json::Value Data,
+                          llvm::BumpPtrAllocator &Allocator,
+                          DenseMap<StringRef, ASTNode *> &Partials,
+                          DenseMap<StringRef, Lambda> &Lambdas,
+                          DenseMap<StringRef, SectionLambda> &SectionLambdas,
+                          DenseMap<char, StringRef> &Escapes);
+
+private:
+  llvm::json::Value findContext();
+  Type T;
+  SmallString<128> RawBody;
+  SmallString<128> Body;
+  ASTNode *Parent;
+  std::vector<ASTNode *> Children;
+  const Accessor Accessor;
+  llvm::json::Value LocalContext;
+};
+
+// A Template represents the container for the AST and the partials
+// and Lambdas that are registered with it.
+class Template {
+public:
+  Template(StringRef TemplateStr);
+
+  SmallString<128> render(llvm::json::Value Data);
+
+  void registerPartial(StringRef Name, StringRef Partial);
+
+  void registerLambda(StringRef Name, Lambda Lambda);
+
+  void registerLambda(StringRef Name, SectionLambda Lambda);
+
+  // By default the Mustache Spec Specifies that HTML special characters
+  // should be escaped. This function allows the user to specify which
+  // characters should be escaped.
```

Comment:
How useful/necessary is it to support custom escapes? I'm wondering if it would be nicer to make the list of escapes some constexpr function instead.

Reply:
It might be useful in languages where escaping html wouldn't be helpful like Latex or Markdown